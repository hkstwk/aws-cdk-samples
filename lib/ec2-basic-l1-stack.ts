import * as cdk from 'aws-cdk-lib';
import {Construct} from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import {AmazonLinuxCpuType, AmazonLinuxGeneration, AmazonLinuxKernel} from 'aws-cdk-lib/aws-ec2';

export class Ec2BasicL1Stack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        // Get default Vpc from the AWS environment this stack is deployed to
        const defaultVpc = ec2.Vpc.fromLookup(this, 'DefaultVPC', {isDefault: true});

        // Create security group, necessary because the default sg created does not allow inbound traffic
        const mySecurityGroup = new ec2.SecurityGroup(this,'ec2-basic-sg-l1', {
            vpc: defaultVpc,
            securityGroupName: 'ec2-basic-sg-l1',
            allowAllOutbound: true
        });

        mySecurityGroup.addIngressRule(ec2.Peer.anyIpv4(), ec2.Port.tcp(80), 'allow inbound http from anywhere');

        // get the imageId to avoid manual lookup and assignment like 'ami-0fef2f5dd8d0917e8'
        const imageId= new ec2.AmazonLinuxImage({
            cpuType: AmazonLinuxCpuType.X86_64,
            generation: AmazonLinuxGeneration.AMAZON_LINUX_2023,
            kernel: AmazonLinuxKernel.KERNEL6_1,
        }).getImage(this).imageId;

        // Create ec2 t2 micro instance using L1 cfn construct (all options configurable)
        new ec2.CfnInstance(this, 'ec2-CfnInstance', {
            keyName: undefined,
            imageId: imageId,
            instanceType: ec2.InstanceType.of(ec2.InstanceClass.T2, ec2.InstanceSize.MICRO).toString(),
            securityGroupIds: [ mySecurityGroup.securityGroupId ],
            privateDnsNameOptions: {
                enableResourceNameDnsAaaaRecord: false,
                enableResourceNameDnsARecord: true,
                hostnameType: 'ip-name',
            },
            tags: [ { key: 'description', value: 'test aws-cdk'}],
            userData: btoa('#!/bin/bash\n' +
                '# Use this for your user data (script from top to bottom)\n' +
                '# install httpd (Linux 2 version)\n' +
                'yum update -y\n' +
                'yum install -y httpd\n' +
                'systemctl start httpd\n' +
                'systemctl enable httpd\n' +
                'echo "<h1>Hello World from $(hostname -f)</h1>" > /var/www/html/index.html')
        });
    }
}


