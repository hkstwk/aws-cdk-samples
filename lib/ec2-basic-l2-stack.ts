import * as cdk from 'aws-cdk-lib';
import {Construct} from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import {AmazonLinuxCpuType, AmazonLinuxGeneration, AmazonLinuxKernel, UserData} from 'aws-cdk-lib/aws-ec2';

export class Ec2BasicL2Stack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        // Get default Vpc from the AWS environment this stack is deployed to
        const defaultVpc = ec2.Vpc.fromLookup(this, 'DefaultVPC', {isDefault: true});

        // Create security group, necessary because the default sg created does not allow inbound traffic
        const mySecurityGroup = new ec2.SecurityGroup(this,'ec2-basic-sg-l2', {
            vpc: defaultVpc,
            securityGroupName: 'ec2-basic-sg-l2',
            allowAllOutbound: true
        });
        mySecurityGroup.addIngressRule(ec2.Peer.anyIpv4(), ec2.Port.tcp(80), 'allow inbound http from anywhere');

        // Create ec2 t2 micro instance using L2 construct
        new ec2.Instance(this, 'basic-ec2-instance', {
            instanceName: 'cdk-basic-ec2-instance',
            keyPair: undefined,
            vpc: defaultVpc,
            machineImage: new ec2.AmazonLinuxImage({
                cpuType: AmazonLinuxCpuType.X86_64,
                generation: AmazonLinuxGeneration.AMAZON_LINUX_2023,
                kernel: AmazonLinuxKernel.KERNEL6_1,
            }),
            instanceType: ec2.InstanceType.of(ec2.InstanceClass.T2, ec2.InstanceSize.MICRO),
            securityGroup: mySecurityGroup,
            requireImdsv2: true,
            ssmSessionPermissions: false,
            userData: UserData.custom('#!/bin/bash\n' +
                '# Use this for your user data (script from top to bottom)\n' +
                '# install httpd (Linux 2 version)\n' +
                'yum update -y\n' +
                'yum install -y httpd\n' +
                'systemctl start httpd\n' +
                'systemctl enable httpd\n' +
                'echo "<h1>Hello World from $(hostname -f)</h1>" > /var/www/html/index.html')
        });
    }
}
