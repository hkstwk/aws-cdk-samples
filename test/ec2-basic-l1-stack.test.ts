import * as cdk from 'aws-cdk-lib';
import {Template} from "aws-cdk-lib/assertions";
import {Ec2BasicL1Stack} from "../lib/ec2-basic-l1-stack";

let app: cdk.App, stack: Ec2BasicL1Stack, template: Template;

beforeAll(() => {
    app = new cdk.App();
    stack = new Ec2BasicL1Stack(app, 'Ec2BasicL1TestStack', {
        env: {
            account: '123456789012',
            region: 'eu-west-2'
        }
    });
    template = Template.fromStack(stack);
    console.dir(template, {depth: null});
});

describe('Resource count tests', () => {
    it('should have one SecurityGroup ', () => {
        template.resourceCountIs('AWS::EC2::SecurityGroup', 1);
    })
    it('should have one EC2 Instance ', () => {
        template.resourceCountIs('AWS::EC2::Instance', 1);
    })
})

describe('SecurityGroup tests', () => {
    it('should have correct groupname', () => {
        template.hasResourceProperties('AWS::EC2::SecurityGroup', {GroupName: 'ec2-basic-sg-l1'})
    });
    it('should have correct outbound rules', () => {
        template.hasResourceProperties('AWS::EC2::SecurityGroup', {
            SecurityGroupEgress: [{
                CidrIp: '0.0.0.0/0',
                IpProtocol: '-1'
            }]
        })
    });
})

    describe('Ec2Instance tests', () => {
        it('should have correct array of security groups', () => {
            template.hasResourceProperties('AWS::EC2::Instance', {SecurityGroupIds: [{'Fn::GetAtt': ['ec2basicsgl165B85E91', 'GroupId']}]});
        })
        it('should have correct instance type', () => {
            template.hasResourceProperties('AWS::EC2::Instance', {InstanceType: "t2.micro"})
        })
        it('should have correct PrivateDnsNameOptions', () => {
            template.hasResourceProperties('AWS::EC2::Instance', {
                PrivateDnsNameOptions: {
                    HostnameType: 'ip-name',
                    EnableResourceNameDnsARecord: true,
                    EnableResourceNameDnsAAAARecord: false
                }
            })
        })
    })

    describe("Ec2BasicL1Stack", () => {
        test("matches the snapshot", () => {
            expect(template.toJSON()).toMatchSnapshot();
        });
    });
