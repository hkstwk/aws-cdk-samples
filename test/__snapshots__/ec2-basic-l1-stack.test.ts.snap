// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ec2BasicL1Stack matches the snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SsmParameterValueawsserviceamiamazonlinuxlatestal2023amikernel61x8664C96584B6F00A464EAD1953AFF4B05118Parameter": {
      "Default": "/aws/service/ami-amazon-linux-latest/al2023-ami-kernel-6.1-x86_64",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>",
    },
  },
  "Resources": {
    "ec2CfnInstance": {
      "Properties": {
        "ImageId": {
          "Ref": "SsmParameterValueawsserviceamiamazonlinuxlatestal2023amikernel61x8664C96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "InstanceType": "t2.micro",
        "PrivateDnsNameOptions": {
          "EnableResourceNameDnsAAAARecord": false,
          "EnableResourceNameDnsARecord": true,
          "HostnameType": "ip-name",
        },
        "SecurityGroupIds": [
          {
            "Fn::GetAtt": [
              "ec2basicsgl165B85E91",
              "GroupId",
            ],
          },
        ],
        "Tags": [
          {
            "Key": "description",
            "Value": "test aws-cdk",
          },
        ],
        "UserData": "IyEvYmluL2Jhc2gKIyBVc2UgdGhpcyBmb3IgeW91ciB1c2VyIGRhdGEgKHNjcmlwdCBmcm9tIHRvcCB0byBib3R0b20pCiMgaW5zdGFsbCBodHRwZCAoTGludXggMiB2ZXJzaW9uKQp5dW0gdXBkYXRlIC15Cnl1bSBpbnN0YWxsIC15IGh0dHBkCnN5c3RlbWN0bCBzdGFydCBodHRwZApzeXN0ZW1jdGwgZW5hYmxlIGh0dHBkCmVjaG8gIjxoMT5IZWxsbyBXb3JsZCBmcm9tICQoaG9zdG5hbWUgLWYpPC9oMT4iID4gL3Zhci93d3cvaHRtbC9pbmRleC5odG1s",
      },
      "Type": "AWS::EC2::Instance",
    },
    "ec2basicsgl165B85E91": {
      "Properties": {
        "GroupDescription": "Ec2BasicL1TestStack/ec2-basic-sg-l1",
        "GroupName": "ec2-basic-sg-l1",
        "SecurityGroupEgress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "allow inbound http from anywhere",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
        ],
        "VpcId": "vpc-12345",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
